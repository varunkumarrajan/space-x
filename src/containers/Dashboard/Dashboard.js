// React
import React from 'react';
import 'react-block-ui/style.css';
// Labels
import labels from 'config/Labels';
// Axios config
import axiosConfig from 'config/Axios.config';
// Config
import config from 'config/Config';
// Constant
import constant from 'constant/Constant';
// Components
import Filters from 'components/Filters/Filters';
import Cards from 'components/Cards/Cards';

// Destructuring
const { noContent } = labels;
const {
  apiURL: { launchQuery },
  filterDetails,
  filterTypes,
} = config;
const { ERROR_CONTENT, DELAY } = constant;

/**
 * @description Add filters & content component and view the dashboard
 */
const Dashboard = () => {
  // Local states
  const [launchDetails, setLaunchDetails] = React.useState([]);
  const [error, setError] = React.useState('');
  const [loading, setLoading] = React.useState(false);
  const [filters, setFilters] = React.useState(
    JSON.parse(JSON.stringify(filterTypes))
  );

  // Call api on the first time
  React.useEffect(() => {
    // Call the api
    handleApiCall();
  }, []);

  // Call api when filters update
  React.useEffect(() => {
    // Call the api
    handleApiCall(...[...filters.map((filter) => filter.value)]);
  }, [filters]);

  /**
   * @description set the error message into error state
   * @param {string} error
   */
  const handleError = (error) => {
    // Set Error
    setError(error);
    // Put timeout for clear the error
    setTimeout(() => setError(''), DELAY);
    // Stop Loading
    setLoading(false);
  };

  /**
   * @description call the api
   * @param {string} [launch='']
   * @param {string} [land='']
   * @param {string} [year='']
   */
  const handleApiCall = (year = '', launch = '', land = '') => {
    // Start Loading
    setLoading(true);
    try {
      // Hit http call throw axios
      axiosConfig.get(launchQuery(launch, land, year)).then((lDetails) => {
        // Check data present and it's type of array
        if (lDetails?.data && Array.isArray(lDetails?.data)) {
          // Update the launch details state
          setLaunchDetails(lDetails.data);
          // Stop Loading
          setLoading(false);
        } else {
          // Set Error
          handleError(ERROR_CONTENT);
        }
      });
    } catch (e) {
      // Set Error
      handleError(e.toString());
    }
  };

  /**
   * @description update the filter values
   * @param {string} filter
   * @param {string} filterType
   */
  const handleFilter = (filter, filterType) =>
    setFilters(
      JSON.parse(JSON.stringify(filters)).map((f) =>
        f.key === filterType ? { ...f, value: filter } : f
      )
    );

  // Return html
  return (
    <React.Fragment>
      {error && <h4>{error}</h4>}
      <Filters
        content={{
          title: labels.filters.title,
          filterDetails,
          filterTypes,
          filters,
        }}
        handleFilter={handleFilter}
      />
      <Cards
        content={{
          loading,
          launchDetails,
          labels: labels.filters,
          noContent,
        }}
      />
    </React.Fragment>
  );
};

// Default dashboard
export default Dashboard;
