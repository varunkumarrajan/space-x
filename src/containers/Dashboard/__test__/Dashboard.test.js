// React
import React from 'react';
// Testing library
import {  cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';

import Dashboard from './../Dashboard';

afterEach(cleanup);

describe('Dashboard', () => {

  it('Dashboard Snapshot', () => {
    const dashboard = renderer.create(<Dashboard />).toJSON();
    expect(dashboard).toMatchSnapshot();
  });
});
