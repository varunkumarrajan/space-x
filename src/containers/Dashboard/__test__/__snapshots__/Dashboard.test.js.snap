// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Dashboard Snapshot 1`] = `
Array [
  "",
  <div
    className="sc-gsTCUz eDbDeT"
  >
    <section
      className="sc-dlfnbm jsrkcH"
    >
      <h4
        className="sc-hKgILt efvMBG"
      >
        Filters
      </h4>
      <div
        className="sc-eCssSg eidbcA"
      >
        Launch Year
      </div>
      <div
        className="sc-jSgupP hUUCHt"
      >
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2006
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2007
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2008
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2009
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2010
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2011
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2012
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2013
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2014
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2015
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2016
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2017
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2018
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2019
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          2020
        </button>
      </div>
      <div
        className="sc-eCssSg eidbcA"
      >
        Successful Launch
      </div>
      <div
        className="sc-jSgupP hUUCHt"
      >
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          True
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          False
        </button>
      </div>
      <div
        className="sc-eCssSg eidbcA"
      >
        Successful Landing
      </div>
      <div
        className="sc-jSgupP hUUCHt"
      >
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          True
        </button>
        <button
          className="sc-bdfBwQ hbzmUa"
          data-testid="buttonId"
          onClick={[Function]}
        >
          False
        </button>
      </div>
    </section>
  </div>,
  <div
    aria-busy={false}
  >
    <section
      className="sc-dQppl elOJqg"
    >
      <h2
        className="sc-bqyKva hXgQog"
      >
        No Records Found
      </h2>
    </section>
    <span />
  </div>,
]
`;
